lang/lang.go

```diff
@@ -0,0 +1,185 @@
+	return LocalizeKey(in, in, data...)
+}
+
+// LocalizeKey asks the translation engine for the translation with specific ID.
+// If it cannot be found then the fallback will be used.
+// The string can be templated and the template data can be passed as a struct with exported fields,
+// or as a map of string keys to any suitable value.
+func LocalizeKey(key, fallback string, data ...any) string {
+	var d0 any
+	if len(data) > 0 {
+		d0 = data[0]
+	}
+
+	ret, err := localizer.Localize(&i18n.LocalizeConfig{
+		DefaultMessage: &i18n.Message{
+			ID:    key,
+			Other: fallback,
+		},
+		TemplateData: d0,
+	})
+
+	if err != nil {
+		fyne.LogError("Translation failure", err)
+		return fallbackWithData(key, fallback, d0)
+	}
+	return ret
+}
+
+// LocalizePlural asks the translation engine to translate a string from one of a number of plural forms.
+// This behaves like the ngettext function, with the `count` parameter determining the plurality looked up.
+// The string can be templated and the template data can be passed as a struct with exported fields,
+// or as a map of string keys to any suitable value.
+func LocalizePlural(in string, count int, data ...any) string {
+	return LocalizePluralKey(in, in, count, data...)
+}
+
+// LocalizePluralKey asks the translation engine for the translation with specific ID in plural form.
+// This behaves like the npgettext function, with the `count` parameter determining the plurality looked up.
+// If it cannot be found then the fallback will be used.
+// The string can be templated and the template data can be passed as a struct with exported fields,
+// or as a map of string keys to any suitable value.
+func LocalizePluralKey(key, fallback string, count int, data ...any) string {
+	var d0 any
+	if len(data) > 0 {
+		d0 = data[0]
+	}
+
+	ret, err := localizer.Localize(&i18n.LocalizeConfig{
+		DefaultMessage: &i18n.Message{
+			ID:    key,
+			Other: fallback,
+		},
+		PluralCount:  count,
+		TemplateData: d0,
+	})
+
+	if err != nil {
+		fyne.LogError("Translation failure", err)
+		return fallbackWithData(key, fallback, d0)
+	}
+	return ret
+}
+
+// AddTranslations allows an app to load a bundle of translations.
+// The language that this relates to will be inferred from the resource name, for example "fr.json".
+// The data should be in json format.
+func AddTranslations(r fyne.Resource) error {
+	return addLanguage(r.Content(), r.Name())
+}
+
+// AddTranslationsForLocale allows an app to load a bundle of translations for a specified locale.
+// The data should be in json format.
+func AddTranslationsForLocale(data []byte, l fyne.Locale) error {
+	return addLanguage(data, l.String()+".json")
+}
+
+func addLanguage(data []byte, name string) error {
+	_, err := bundle.ParseMessageFileBytes(data, name)
+	return err
+}
+
+func init() {
+	bundle = i18n.NewBundle(language.English)
+	bundle.RegisterUnmarshalFunc("json", json.Unmarshal)
+
+	loadTranslationsFromFS(translations, "translations")
```

Comment:
Should we really parse all translations for all languages on startup? This will not only slow down startup with many languages but also add extra memory pressure. Would it not be better to only load the requested language?

Reply:
How will you know which language files are needed?
If the user reports a Brazilian locale we need to load the Portuguese language, for example (if nothing exists for Brazil). 
Similarly Canadian French locale is an English language country with a French language pack. The lookup code will do this correctly, but only if we have loaded the languages needed.

If performance is an issue we could split the parse into two, loading all the files first to do the lookup and then parsing the messages only from those needed. However we must also ensure that the fallback path is honoured, so that an "en_GB" and "en" are not assumed to be the same and the more specific could provide only the messages which differ.